iree/samples/simple_embedding/device_vmvx_sync.c

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// A example of setting up the vmvx-sync driver.
+
+#include <stddef.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+#include "iree/hal/local/loaders/vmvx_module_loader.h"
+#include "iree/hal/local/sync_device.h"
+
+// Compiled module embedded here to avoid file IO:
+#include "iree/samples/simple_embedding/simple_embedding_test_bytecode_module_vmvx_c.h"
+
+iree_status_t create_sample_device(iree_hal_device_t** device) {
+  // Set parameters for the device created in the next step.
+  iree_hal_sync_device_params_t params;
+  iree_hal_sync_device_params_initialize(&params);
+
+  iree_vm_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance));
+
+  iree_hal_executable_loader_t* loader = NULL;
+  IREE_RETURN_IF_ERROR(iree_hal_vmvx_module_loader_create(
+      instance, iree_allocator_system(), &loader));
+
+  iree_string_view_t identifier = iree_make_cstring_view("vmvx");
+
+  // Create the synchronous device and release the loader afterwards.
+  IREE_RETURN_IF_ERROR(
```

Comment:
there's a few leaks in this function (the instance is always leaked and the loader is leaked if device creation fails)
you can restructure this with an if(ok) pattern:

```c
  // instance is fine to RETURN_IF_ERROR because nothing would leak)

  iree_hal_executable_loader_t* loader = NULL;
  iree_status_t status = iree_hal_vmvx_module_loader_create(
      instance, iree_allocator_system(), &loader);
  iree_vm_instance_release(instance);

  if (iree_status_is_ok(status)) {
    status = iree_hal_sync_device_create(...)
  }

  iree_hal_executable_loader_release(loader);
  return status;
```

Reply:
Thanks for catching this. I will also fix the leaks of other devices in this sample.